common/src/main/scala/explore/components/graphql/SubscriptionRenderMod.scala

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2016-2020 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package explore.components.graphql
+
+import japgolly.scalajs.react._
+import japgolly.scalajs.react.vdom.html_<^._
+import react.common._
+import react.semanticui.elements.icon.Icon
+import react.semanticui.sizes._
+import clue.GraphQLStreamingClient
+import cats.implicits._
+import cats.effect.IO
+import crystal._
+import crystal.react.io.implicits._
+import diode.data._
+import diode.react.ReactPot._
+import cats.effect.ContextShift
+import crystal.react.StreamRendererMod
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import cats.effect.Timer
+
+final case class SubscriptionRenderMod[D, A](
+  subscribe:      IO[GraphQLStreamingClient[IO]#Subscription[D]],
+  streamModifier: fs2.Stream[IO, D] => fs2.Stream[IO, A] = identity[fs2.Stream[IO, D]] _
+)(
+  val valueRender: (A, StreamRendererMod.ModState[A]) => VdomNode,
+  val onNewData:   IO[Unit] = IO.unit
+)(
+  implicit val cs: ContextShift[IO],
+  val timer:       Timer[IO]
+) extends ReactProps {
+  override def render: VdomElement =
+    SubscriptionRenderMod.component(this.asInstanceOf[SubscriptionRenderMod[Any, Any]])
+}
+
+object SubscriptionRenderMod {
+  protected type Props[D, A] = SubscriptionRenderMod[D, A]
+
+  protected final case class State[D, A](
+    subscription: Option[
+      GraphQLStreamingClient[IO]#Subscription[D]
+    ] = None,
+    renderer: Option[StreamRendererMod[Pot[A]]] = None
+  )
+
+  implicit protected def propsReuse[D, A]: Reusability[Props[D, A]] = Reusability.always
+  implicit protected def stateReuse[D, A]: Reusability[State[D, A]] = Reusability.never
+
+  protected def componentBuilder[D, A] =
+    ScalaComponent
+      .builder[Props[D, A]]("SubscriptionRenderMod")
+      .initialState(State[D, A]())
+      .render { $ =>
+        <.div(
+          $.state.renderer.whenDefined(
+            _ { (resultsPot, modState) =>
+              <.div(
+                resultsPot.renderPending(_ => Icon(name = "spinner", loading = true, size = Large)),
+                resultsPot.render(results => $.props.valueRender(results, f => modState(_.map(f))))
+              )
+            }
+          )
+        )
+      }
+      .componentWillMount { $ =>
+        $.props.subscribe
+          .flatMap { subscription =>
+            implicit val cs    = $.props.cs
+            implicit val timer = $.props.timer
+
+            $.modStateIO(_ =>
+              State(
+                subscription.some,
+                StreamRendererMod
+                  .build(
+                    $.props
+                      .streamModifier(subscription.stream)
+                      .map[Pot[A]](a => Ready(a))
+                      .flatTap(_ => fs2.Stream.eval($.props.onNewData))
+                      .cons1(Pending()),
+                    holdAfterMod = (2 seconds).some
```

Comment:
What's this?

Reply:
That's the mechanism to avoid flicker.

Crystal's `StreamRendererMod` is a stateful component bound to a `Stream`, whose state is updated either by:
- receiving a new value in the `Stream`
- executing the `modState` effect that it provides to children.

When `modState` is executed repeatedly and the modification sent to the server, then flicker would happen when the server sent intermediate values in the stream. To avoid this, `StreamRendererMod` has this optional "hold" mechanism in which it will not pass stream values to children during some time after receiving a `modState`.

The amount of time it should wait before re-enabling the stream is that parameter `holdAfterMod`.

During the hold, the last value received in the stream is remembered and it is passed down the line when the hold ends. If another `modState` is received during the hold, the timer is reset.